local-cluster/tests/local_cluster.rs

```diff
@@ -3493,146 +3502,200 @@ fn test_fork_choice_refresh_old_votes() {
+                            false,
+                        );
 
-            // Restart the smallest validator that we killed earlier in `on_partition_start()`
-            drop(smallest_blockstore);
-            cluster.restart_node(
-                &context.smallest_validator_key,
-                context.alive_stake3_info.take().unwrap(),
-                SocketAddrSpace::Unspecified,
-            );
+                        // Also copy all the blocks from the heavier partition so the smallest validator will
+                        // not vote again on its own fork
+                        copy_blocks(
+                            heaviest_fork_latest_vote,
+                            &heaviest_blockstore,
+                            &smallest_blockstore,
+                            false,
+                        );
 
-            loop {
-                // Wait for node to vote on the first slot on the less heavy fork, so it'll need
-                // a switch proof to flip to the other fork.
-                // However, this vote won't land because it's using an expired blockhash. The
-                // fork structure will look something like this after the vote:
-                /*
-                     1 (2%, killed and restarted) --- 200 (37%, lighter fork)
-                    /
-                    0
-                    \-------- 4 (38%, heavier fork)
-                */
-                if let Some((last_vote_slot, _last_vote_hash)) =
-                    last_vote_in_tower(&smallest_ledger_path, &context.smallest_validator_key)
-                {
-                    // Check that the heaviest validator on the other fork doesn't have this slot,
-                    // this must mean we voted on a unique slot on this fork
-                    if last_vote_slot == first_slot_in_lighter_partition {
-                        info!(
-                            "Saw vote on first slot in lighter partition {}",
-                            first_slot_in_lighter_partition
+                        // Simulate a vote for the `first_slot_in_lighter_partition`
+                        let bank_hash = lighter_fork_blockstore
+                            .get_bank_hash(context.first_slot_in_lighter_partition)
+                            .unwrap();
+                        cluster_tests::apply_votes_to_tower(
+                            &context
+                                .smallest_validator_info
+                                .as_ref()
+                                .unwrap()
+                                .info
+                                .keypair,
+                            vec![(context.first_slot_in_lighter_partition, bank_hash)],
+                            smallest_ledger_path,
                         );
+
+                        // Restart the smallest validator that we killed earlier in `on_partition_start()`
+                        drop(smallest_blockstore);
```

Comment:
Can we move line 3653~3654 to just before restart_node so they stay together? We can explain here that we are ending the loop because the wait is over?

Reply:
oh yeah that comment is in the wrong place, fixed